src/patch.rkt

```diff
@@ -0,0 +1,318 @@
+#lang racket
+
+(require "core/matcher.rkt" "core/taylor.rkt" "core/simplify.rkt"
+         "alternative.rkt" "common.rkt" "interface.rkt" "programs.rkt"
+         "timeline.rkt" "syntax/rules.rkt" "syntax/sugar.rkt")
+
+(provide
+  (contract-out
+   [patch-table-add! (-> expr? (listof symbol?) boolean? void?)]
+   [patch-table-get (-> expr? (listof alt?))]
+   [patch-table-has-expr? (-> expr? boolean?)]
+   [patch-table-runnable? (-> boolean?)]
+   [patch-table-run (-> (listof alt?))]
+   [patch-table-run-simplify (-> (listof alt?) (listof alt?))]))
```

Comment:
Why are all of these public, like `patch-table-add!`? Shouldn't that be merely internal?

Reply:
The mainloop calls all of these in separate locations. By public do you mean contracts? I could remove those. It was helpful when debugging.

Reply:
Huh, ok, I didn't read the mainloop code accurately enough...

Reply:
Ok, some follow-up questions.

1. Why not generate cached even when not `patch-table-runnable?`? Or put another way, why not eliminate `patch-table-runnable?` and just make `patch-table-run` be a no-op if there's nothing queued?
2. `patch-table-run-simplify` is kind of ugly but I do understand why it's there. I wonder if initial simplify matters any?
3. Ok, `patch-table-get` is just used to generate cached things. What if that was done transparently by `patch-table-run`? That is, what if `patch-table-run` was responsible for checking what's in the cache and in that case only using the cache?

Reply:
The upshot of question 3 is—why not store the queue in the mainloop, where it is already sort of duplicated by `^locs^` and `^lowlocs^`, and just expose `patch-table-run` from `patch.rkt`?

Reply:
I like (1). Hmm (3*) is interesting. That actually makes more sense.

Reply:
Let's try it!